### PR TITLE
add event payload cryptographer for personal data

### PR DESCRIPTION
**Pull request by DavidBadura**
I took this blog post as inspiration and built a prototype:

https://www.eventstore.com/blog/protecting-sensitive-data-in-event-sourced-systems-with-crypto-shredding-1

This is what an event looks like when you use personal data:

```php
<?php
use Patchlevel\EventSourcing\Attribute\DataSubjectId;
use Patchlevel\EventSourcing\Attribute\Event;
use Patchlevel\EventSourcing\Attribute\PersonalData;
use Patchlevel\EventSourcing\Serializer\Normalizer\IdNormalizer;

#[Event('profile.created')]
final class ProfileCreated
{
    public function __construct(
        #[IdNormalizer]
        #[DataSubjectId]
        public ProfileId $profileId,
        #[PersonalData(fallback: 'unknown')]
        public string $name,
    ) {
    }
}
```

And you can use a processor to remove personal data:

```php
use Patchlevel\EventSourcing\Attribute\Processor;
use Patchlevel\EventSourcing\Attribute\Subscribe;
use Patchlevel\EventSourcing\Cryptography\Store\CipherKeyStore;
use Patchlevel\EventSourcing\Message\Message;

#[Processor('delete_personal_data')]
final class DeletePersonalDataProcessor
{
    public function __construct(
        private readonly CipherKeyStore $cipherKeyStore,
    ) {
    }

    #[Subscribe(UserHasRequestedDeletion::class)]
    public function handleUserHasRequestedDeletion(Message $message): void
    {
        $event = $message->event();

        $this->cipherKeyStore->remove($event->profileId->toString());
    }
}
```

### CODE REVIEW

File: src/Attribute/DataSubjectId.php [hunk 1]
```diff
@@ -0,0 +1,12 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Patchlevel\EventSourcing\Attribute;
+
+use Attribute;
+
+#[Attribute(Attribute::TARGET_PROPERTY)]
+final class DataSubjectId
```
Comment (DanielBadura) [inline src/Attribute/DataSubjectId.php:10]: This naming coming from the article you mentioned at the PR. But i could find in the article *why* it is named like it is. For me the naming is not really making sense in the context. Do I miss something?

File: src/Cryptography/Store/InMemoryCipherKeyStore.php [hunk 2]
```diff
@@ -0,0 +1,33 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Patchlevel\EventSourcing\Cryptography\Store;
+
+use Patchlevel\EventSourcing\Cryptography\Cipher\CipherKey;
+
+final class InMemoryCipherKeyStore implements CipherKeyStore
```
Comment (DanielBadura) [inline src/Cryptography/Store/InMemoryCipherKeyStore.php:9]: Im not sure if this makes sense to have in our `src` dir. For testing purposes i can understand that, but on production systems?

File: src/Cryptography/Store/DoctrineCipherKeyStore.php [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+ *     subject_id: non-empty-string,
+ *     crypto_key: non-empty-string,
+ *     crypto_method: non-empty-string,
+ *     crypto_iv: non-empty-string
+ * }
+ */
+final class DoctrineCipherKeyStore implements CipherKeyStore, DoctrineSchemaConfigurator
+{
+    /** @var array<string, CipherKey> */
+    private array $keys = [];
```
Comment (DanielBadura) [inline src/Cryptography/Store/DoctrineCipherKeyStore.php]: maybe name it keyCache or sth like that?

File: src/Attribute/DataSubjectId.php (hunk 1 above)
Comment (DavidBadura) [reply src/Attribute/DataSubjectId.php:10]: The point is that every person-related data must be assigned to a subject. A subject can be a person. To refer to this subject, you need to use a identifier. So we have the name `SubjectId`. And I think `Data` is a prefix to scope it and understand that this is about a subject ID related to the data.

File: src/Cryptography/Store/InMemoryCipherKeyStore.php (hunk 2 above)
Comment (DavidBadura) [reply src/Cryptography/Store/InMemoryCipherKeyStore.php:9]: Classes that are used (only) for tests and offered to the developer also belong in the src folder.
I definitely want to test this class too. This is not a stub or mock, but a valid implementation of the interface.

**Review by DanielBadura — APPROVED** [review]

**Pull request merged**